### ISSUE TITLE
Analysis results are non-deterministic

### ISSUE THREAD

**Issue by lnicola** [rust-analyzer/rust-analyzer#1552]
```
$ cargo run --release -p ra_cli -- analysis-stats
    Finished release [optimized + debuginfo] target(s) in 0.06s
     Running `target/release/ra_cli analysis-stats`
Database loaded, 256 roots, 185.744807ms
Crates in this dir: 28
Total modules found: 285
Total declarations: 9644
Total functions: 3253
Total expressions: 64916
Expressions of unknown type: 9217 (14%)
Expressions of partially unknown type: 3414 (5%)
Analysis: 17.890922846s, 0b allocated 0b resident

$ cargo run --release -p ra_cli -- analysis-stats
    Finished release [optimized + debuginfo] target(s) in 0.05s
     Running `target/release/ra_cli analysis-stats`
Database loaded, 256 roots, 184.273592ms
Crates in this dir: 28
Total modules found: 285
Total declarations: 9644
Total functions: 3253
Total expressions: 64916
Expressions of unknown type: 9218 (14%)
Expressions of partially unknown type: 3410 (5%)
Analysis: 17.724273571s, 0b allocated 0b resident
```

**Comment by bjorn3** [comment]
`HashMap` or `HashSet` iteration?

**Comment by lnicola** [comment]
I haven't bisected it, but this seems fixed as of `0492b18e`.

**Comment by lnicola** [comment]
This seems.. fixed? I'm on `2c735ed734be9b9041921478e0049fffd7160f78` and I got the same results on 10 consecutive runs (note that the toolchain version still matters):

```
Database loaded:     639.53ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.17s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.86s, 124ginstr
Total:               21.03s, 190ginstr
Database loaded:     639.11ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.17s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.94s, 124ginstr
Total:               21.11s, 190ginstr
Database loaded:     484.31ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.15s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.77s, 124ginstr
Total:               20.92s, 190ginstr
Database loaded:     475.05ms, 297minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.22s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.82s, 124ginstr
Total:               21.04s, 190ginstr
Database loaded:     459.08ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.14s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.80s, 124ginstr
Total:               20.95s, 190ginstr
Database loaded:     453.10ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.06s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.72s, 124ginstr
Total:               20.79s, 190ginstr
Database loaded:     474.88ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.15s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.83s, 124ginstr
Total:               20.97s, 190ginstr
Database loaded:     469.32ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.23s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           14.84s, 124ginstr
Total:               21.07s, 190ginstr
Database loaded:     485.15ms, 296minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.21s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           15.02s, 124ginstr
Total:               21.23s, 190ginstr
Database loaded:     487.90ms, 297minstr
  crates: 36, mods: 594, decls: 11461, fns: 8965
Item Collection:     6.23s, 66ginstr
  exprs: 247088, ??ty: 2722 (1%), ?ty: 2106 (0%), !ty: 995                                                                                       
Inference:           15.12s, 124ginstr
Total:               21.35s, 190ginstr
```

Even more, they're the same as on CI:

```
{
    "host": {
        "os": "Ubuntu 18.04.5 LTS",
        "cpu": "Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz",
        "mem": "7121296 kB"
    },
    "timestamp": 1611585595,
    "revision": "2c735ed734be9b9041921478e0049fffd7160f78",
    "metrics": {
        "analysis-stats/ripgrep/total memory": [
            331,
            "MB"
        ],
        "analysis-stats/ripgrep/total time": [
            9949,
            "ms"
        ],
        "analysis-stats/ripgrep/type mismatches": [
            5,
            "#"
        ],
        "analysis-stats/ripgrep/unknown type": [
            143,
            "#"
        ],
        "analysis-stats/self/total memory": [
            1106,
            "MB"
        ],
        "analysis-stats/self/total time": [
            48836,
            "ms"
        ],
        "analysis-stats/self/type mismatches": [
            995,
            "#"
        ],
        "analysis-stats/self/unknown type": [
            2722,
            "#"
        ],
        "analysis-stats/webrender/total memory": [
            597,
            "MB"
        ],
        "analysis-stats/webrender/total time": [
            20928,
            "ms"
        ],
        "analysis-stats/webrender/type mismatches": [
            216,
            "#"
        ],
        "analysis-stats/webrender/unknown type": [
            1503,
            "#"
        ],
        "build": [
            582403,
            "ms"
        ]
    }
}
```

We might still want to check that we don't regress on this, though.
